### PR TITLE
Generate PSA constant names in CMake build dir

### LINKED ISSUE #3365: PSA Crypto: allow to generate PSA constants in another directory
### Description
- Type: Bug and Enhancement
- Priority: Major

---------------------------------------------------------------

In [`rust-psa-crypto`](https://github.com/parallaxsecond/rust-psa-crypto) we are currently building a Rust wrapper around the PSA Crypto interface.
To be able to publish our library on crates.io, it needs to build without modifying the source directory (which is read-only) and only writing to the build directory.

We successfully were able to create the `config.h` file in the build directory with the `--write` option and to compile inside it using CMake.
However during compilation, CMake call the `generate_psa_constants.py` script which creates the `programs/psa/psa_constant_names_generated.c` file inside the source tree. This file is then included with an `#include` macro.

Would it be possible to control the directory where that file is generated? Maybe it should be generated in the CMake build directory by default?

### PR DESCRIPTION
**Pull request by hug-dev**
This commit modifies the generate_psa_constants.py script to take as
input argument the location of where to write the
psa_constant_names_generated.c file.
For make-based build system, this commit does not change anything.
For CMake build system, this commit modifies the generation location of
that file to be inside the build directory and include it from there in
psa_constant_names.c

Fix #3365

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Unfortunately something's going wrong on Windows.
[windows-log.txt.gz](https://github.com/ARMmbed/mbedtls/files/4695300/windows-log.txt.gz)

**Comment by hug-dev** [comment]
On the `.travis.yml` I had to make the following change to pass the CI on Windows:
```diff
- - python scripts/generate_psa_constants.py
+ - python scripts/generate_psa_constants.py programs/psa
```

Maybe something similar is needed on the CI systems that failed? Are the CI scripts for the ones that failed on this repository?

**Comment by ronald-cron-arm** [comment]
Hi Hughes, the error seems to be the following one:
Traceback (most recent call last):
  File "scripts\generate_psa_constants.py", line 401, in <module>
    sys.argv[1] + '/psa_constant_names_generated.c')
IndexError: list index out of range
2020-05-28 09:09:05,194 - VS2013 Release Win32 shipped - ERROR - None
2020-05-28 09:09:05,194 - VS2013 Release Win32 shipped - ERROR - Generating psa constants failed, aborting
Traceback (most recent call last):
  File "windows_testing.py", line 517, in generate_psa_constants
    check=True
  File "C:\Python36\lib\subprocess.py", line 418, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['C:\\Python36\\python.exe', 'scripts\\generate_psa_constants.py']' returned non-zero exit status 1.

**Comment by gilles-peskine-arm** [comment]
The code that calls `generate_psa_constants.py` for the Windows build is indeed in a private repository. That code is agnostic to the version of Mbed TLS, so changing the interface between the two is a bit painful.

Fortunately I think there's an easy solution here that doesn't require any change in the scripts: make the Python script use `argv[1]` if present and default to `programs/psa` otherwise.

**Comment by hug-dev** [comment]
I see. That would be because in the modifications I expect the `generate_psa_constants.py` script to have one argument.

> Fortunately I think there's an easy solution here that doesn't require any change in the scripts: make the Python script use argv[1] if present and default to programs/psa otherwise.

Yes, very good point! Will add that.

File: scripts/generate_psa_constants.py [hunk 1]
```diff
@@ -395,6 +396,8 @@ def generate_psa_constants(header_file_names, output_file_name):
 if __name__ == '__main__':
     if not os.path.isdir('programs') and os.path.isdir('../programs'):
         os.chdir('..')
+    # Allow to change the directory where psa_constant_names_generated.c is written to.
+    OUTPUT_FILE_DIR = sys.argv[1] if len(sys.argv) == 2 else "programs/psa"
```
Comment (ronald-cron-arm) [inline scripts/generate_psa_constants.py:400]: Please document at the begin of the file, the two ways of calling the script: without any argument or with the output file directory and the corresponding two directories where the generated file is going to be stored.

Comment (hug-dev) [reply scripts/generate_psa_constants.py:405]: Done in an added commit!

**Comment by hug-dev** [comment]
Thanks Ronald 😃

File: scripts/generate_psa_constants.py [hunk 2]
```diff
@@ -395,6 +401,8 @@ def generate_psa_constants(header_file_names, output_file_name):
...
```
Comment (gilles-peskine-arm) [inline scripts/generate_psa_constants.py:405]: [PEP8](https://www.python.org/dev/peps/pep-0008/#id45) says all-caps for global constants, lowercase for local variables. Please don't use all-caps for a local variable. (I'm surprised that Pylint didn't complain, it's usually picky about these things.)

Comment (gilles-peskine-arm) [inline scripts/generate_psa_constants.py:405]: `len(sys.argv) >= 2` would make more sense.

Comment (hug-dev) [reply scripts/generate_psa_constants.py:405]: I originally put it lower case but the [CI compained](https://travis-ci.org/github/ARMmbed/mbedtls/jobs/692210799#L1646):
```
scripts/generate_psa_constants.py:400:4: C0103: Constant name "output_file_dir" doesn't conform to UPPER_CASE naming style (invalid-name)
```
This variable is not declared under a function so maybe it is still a global variable?

Comment (gilles-peskine-arm) [reply scripts/generate_psa_constants.py:405]: Oh, I see, it isn't in a function, there's no `main` function in this script. Ok, let's keep it like this for now.

**Comment by hug-dev** [comment]
Thank you!!

**Comment by hug-dev** [comment]
Could the next tagged version of Mbed TLS contain that patch? That would allow us to point to that version instead of an arbitrary commit hash.

**Comment by ronald-cron-arm** [comment]
This has been merged in the development branch thus will be part of the next release, 2.23 I think,

**Comment by hug-dev** [comment]
Great, thanks 😃 !

**Pull request merged**